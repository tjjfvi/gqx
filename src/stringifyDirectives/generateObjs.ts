
import { Obj } from ".";
import { Context } from "..";

export default (ctx: Context) =>
  Object.entries(ctx.objectTypes).map(([type, fields]): Obj => ({
    type,
    props: fields.map(f => ({
      type,
      prop: f.name.value,
      dirs: (f.directives || []).map(d => ({
        type,
        prop: f.name.value,
        dirName: d.name.value,
        dirArgs: d.arguments?.slice(),
      }))
    })),
  }))
