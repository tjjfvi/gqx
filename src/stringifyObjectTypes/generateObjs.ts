
import { Context, Obj, Id, Prop } from ".";
import unwrapType from "../unwrapType";

export default (ctx: Context): [Obj[], Id[], Context] => {
  const ids: Id[] = [];
  const objs: Obj[] =
    Object.keys(ctx.objectTypes)
      .sort()
      .map(name => {
        const { fields, unions, implements: impls } = ctx.objectTypes[name];
        const props = fields.map(field => {
          const id: Id = {
            type: name,
            prop: field.name.value,
          };
          ids.push(id);
          const [{ name: { value: typeName } }, wrap,, wrapHKT] = unwrapType(field.type);
          const shallow = !(typeName in ctx.objectTypes);
          let prop: Prop = {
            id,
            type: typeName,
            shallow,
            wrap,
            wrapHKT,
            args: field.arguments?.slice() ?? [],
            directives: field.directives?.slice() ?? [],
            loc: field.loc,
          };
          return prop;
        }).sort((p, q) => p.id.prop > q.id.prop ? 1 : -1);
        const shallowProps = props.filter(p => p.shallow);
        const deepProps = props.filter(p => !p.shallow);
        const obj: Obj = {
          type: name,
          shallowProps,
          deepProps,
          props,
          unions,
          implements: impls,
          isBase: ctx.baseTypes.includes(name),
        };
        return obj;
      });
  return [objs, ids, ctx];
}
