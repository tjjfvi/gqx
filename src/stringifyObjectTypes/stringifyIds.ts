
import { Id } from ".";
import stringifyId from "./stringifyId";
import indent from "../indent";

const str = (x: string) => JSON.stringify(x);

export default (ids: Id[]) =>
  ids.map(id =>
    [
      `export class ${stringifyId(id)} {`,
      indent`private static _: any;`,
      indent`static type = ${str(id.type)} as const;`,
      indent`static prop = ${str(id.prop)} as const;`,
      indent`static f = ${stringifyId(id)};`,
      `}`,
    ].join("\n")
  ).join("\n")
