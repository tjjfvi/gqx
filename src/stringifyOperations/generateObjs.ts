
import { Context, Obj, Id, Prop } from ".";
import unwrapType from "../unwrapType";

export default (ctx: Context): [Obj[], Context] => {
  const objs: Obj[] =
    ctx.operations
      .map(([prop, name]) => {
        const { fields } = ctx.objectTypes[name];
        const props = fields.map(field => {
          const id: Id = {
            type: name,
            prop: field.name.value,
          };
          const [{ name: { value: typeName } }, wrap] = unwrapType(field.type);
          let prop: Prop = {
            id,
            type: typeName,
            wrap,
            args: field.arguments?.slice() ?? [],
          };
          return prop;
        }).sort((p, q) => p.id.prop > q.id.prop ? 1 : -1);
        const obj: Obj = {
          type: name,
          prop,
          props,
        };
        return obj;
      })
  return [objs, ctx];
}
