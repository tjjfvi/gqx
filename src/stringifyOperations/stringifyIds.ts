
import { Id } from ".";
import stringifyId from "./stringifyId";

export default (ids: Id[]) =>
  ids.map(id => `class ${stringifyId(id)} { declare private __nominal__: any; }\n`).join("")
