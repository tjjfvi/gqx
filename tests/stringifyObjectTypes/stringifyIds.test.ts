
import generateObjs from "../../src/stringifyObjectTypes/generateObjs"
import stringifyIds from "../../src/stringifyObjectTypes/stringifyIds";

test("", () => {
  const [, ids] = generateObjs(utils.getOperations(["stringifyObjectTypes"]));
  const code = stringifyIds(ids);
  expect(code).toParse().toMatchSnapshot();
})
