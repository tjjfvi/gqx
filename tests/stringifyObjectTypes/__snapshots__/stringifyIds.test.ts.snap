// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
"export class BinaryTree$node {
  private static _: any;
  static type = \\"BinaryTree\\" as const;
  static prop = \\"node\\" as const;
  static f = BinaryTree$node;
}
export class BinaryTree$left {
  private static _: any;
  static type = \\"BinaryTree\\" as const;
  static prop = \\"left\\" as const;
  static f = BinaryTree$left;
}
export class BinaryTree$right {
  private static _: any;
  static type = \\"BinaryTree\\" as const;
  static prop = \\"right\\" as const;
  static f = BinaryTree$right;
}
export class DeepOnly$a {
  private static _: any;
  static type = \\"DeepOnly\\" as const;
  static prop = \\"a\\" as const;
  static f = DeepOnly$a;
}
export class DeepOnly$b {
  private static _: any;
  static type = \\"DeepOnly\\" as const;
  static prop = \\"b\\" as const;
  static f = DeepOnly$b;
}
export class Recursive$id {
  private static _: any;
  static type = \\"Recursive\\" as const;
  static prop = \\"id\\" as const;
  static f = Recursive$id;
}
export class Recursive$next {
  private static _: any;
  static type = \\"Recursive\\" as const;
  static prop = \\"next\\" as const;
  static f = Recursive$next;
}
export class ShallowOnly$a {
  private static _: any;
  static type = \\"ShallowOnly\\" as const;
  static prop = \\"a\\" as const;
  static f = ShallowOnly$a;
}
export class ShallowOnly$b {
  private static _: any;
  static type = \\"ShallowOnly\\" as const;
  static prop = \\"b\\" as const;
  static f = ShallowOnly$b;
}"
`;
